test/system/src/test/java/io/pravega/test/system/ConsumptionBasedRetentionWithMultipleReaderGroupsTest.java

```diff
@@ -283,4 +238,181 @@ public void multipleSubscriberCBRTest() throws Exception {
+        readingEventsFromStream(3, reader);
+        readingEventsFromStream(3, reader2);
+
+        log.info("{} generating 2nd stream-cuts for {}/{}", READER_GROUP_3, SCOPE_1, STREAM_1);
+        streamCuts = generateStreamCuts(readerGroup, reader);
+        log.info("{} updating its retention stream-cut to {}", READER_GROUP_3, streamCuts);
+        readerGroup.updateRetentionStreamCut(streamCuts);
+
+        log.info("{} generating 2nd stream-cuts for {}/{}", READER_GROUP_4, SCOPE_1, STREAM_2);
+        streamCuts2 = generateStreamCuts(readerGroup2, reader2);
+        log.info("{} updating its retention stream-cut to {}", READER_GROUP_4, streamCuts2);
+        readerGroup2.updateRetentionStreamCut(streamCuts2);
+
+        // Retention set has two stream cut at 0/360, 0/450
+        // READER_GROUP_3 updated stream cut at 0/300, Subscriber lower bound is 0/300
+        // So truncating at stream cut 0/300 which leaves more data in the stream than min limit
+        // The timeout is set to 3 minutes a little longer than the retention period which is set to 2 minutes
+        // in order to confirm that the retention has taken place.
+        // Check to make sure truncation happened at SLB
+        AssertExtensions.assertEventuallyEquals("Truncation did not take place at offset 300.", true, () -> controller.getSegmentsAtTime(
+                        new StreamImpl(SCOPE_1, STREAM_1), 0L).join().values().stream().anyMatch(off -> off == 300),
+                5000, 3 * 60 * 1000L);
+        // Retention set has two stream cuts at 0/450...0/570
+        // READER_GROUP_4 updated stream cut at 0/390, Subscriber lower bound is 0/390
+        // Since 0/450 is the stream cut satisfying min and max bounds and SLB is lower than this.
+        // Truncation should happen at SLB
+        assertEquals(true, controller.getSegmentsAtTime(
+                new StreamImpl(SCOPE_1, STREAM_2), 0L).join().values().stream().anyMatch(off -> off == 390));
+    }
+
+
+    private void writingEventsToStream(int numberOfEvents, EventStreamWriter<String> writer, String scope, String stream) {
+        for (int i = 0; i < numberOfEvents; i++) {
+            log.info("Writing event to {}/{}", scope, stream);
+            writer.writeEvent(SIZE_30_EVENT).join();
+        }
+    }
+
+    private void readingEventsFromStream(int numberOfEvents, EventStreamReader<String> reader) {
+        EventRead<String> read;
+        for (int i = 0; i < numberOfEvents; i++) {
+            read = reader.readNextEvent(READ_TIMEOUT);
+            assertEquals("data of size 30", read.getEvent());
+        }
+    }
+
+    private Map<Stream, StreamCut> generateStreamCuts(ReaderGroup readerGroup, EventStreamReader<String> reader) {
+        CompletableFuture<Map<Stream, StreamCut>> futureCuts = readerGroup.generateStreamCuts(streamCutExecutor);
+        // Wait for 5 seconds to force reader group state update. This will allow for the silent
```

Comment:
This delay is only caused by the imposed delay of refreshing the reader group. If you use the same reader group object that the reader holds (This may require some reflection trickery) then it is possible to add the stream cut, and call read and observe the result immediately.

Reply:
Have removed the delay.